src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+    // Anything otherwise unavailable just has a bump fee of 0
+    for (const auto& outpoint : outpoints) {
+        if (const auto ptx{mempool.GetConflictTx(outpoint)}) {
+            // The conflicting transaction is to-be-replaced
+            to_be_replaced.insert(ptx->GetHash());
+        }
+
+        if (!mempool.exists(GenTxid::Txid(outpoint.hash))) {
+            // This UTXO is either confirmed or not yet submitted to mempool.
+            // In the former case, no bump fee is required.
+            // In the latter case, we have no information, so just return 0.
+            this->bump_fees.emplace(std::make_pair(outpoint, 0));
+        } else {
+            // This UTXO is unconfirmed, in the mempool, and available to spend.
+            auto it = outpoints_needed_by_txid.find(outpoint.hash);
+            if (it != outpoints_needed_by_txid.end()) {
+                it->second.push_back(outpoint);
+            } else {
+                std::vector<COutPoint> outpoints_of_tx({outpoint});
+                outpoints_needed_by_txid.emplace(std::make_pair(outpoint.hash, outpoints_of_tx));
+                // Instead of operating on the entire mempool, just run the mining algorithm on the
+                // cluster of relevant transactions, which we'll store in mapModifiedTx.
+            }
+        }
+    }
+    // Calculate the cluster and construct the entry map.
+    std::vector<uint256> txids_needed;
+    std::transform(outpoints_needed_by_txid.cbegin(),
+                   outpoints_needed_by_txid.cend(),
+                   std::back_inserter(txids_needed),
+                   [](const auto& pair) { return pair.first; });
+    const auto& cluster = mempool.CalculateCluster(txids_needed);
+    for (const auto& txiter : cluster) {
+        if (to_be_replaced.find(txiter->GetTx().GetHash()) == to_be_replaced.end()) {
+            auto [mapiter, success] = entries_by_txid.emplace(std::make_pair(txiter->GetTx().GetHash(), MockMempoolEntry(txiter)));
+            assert(success);
+            entries.push_back(mapiter);
+        } else {
+            auto outpoints_it = outpoints_needed_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != outpoints_needed_by_txid.end()) {
+                for (const auto& outpoint : outpoints_it->second) {
+                    this->bump_fees.emplace(std::make_pair(outpoint, 0));
+                }
+            }
+        }
+    }
+
+    // Remove the to-be-replaced transactions and build the descendant_set_by_txid cache.
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
```

Comment:
898ad9d5904f1b689d18d94f20d92500cf443758 The result of this `find()` could be `entries_by_txid.end()` (not found), because in the previous loop (also over `cluster`), if the transaction is found in `to_be_replaced` (so we take the `else` path), then the transaction is not added to `entries_by_txid`. I think it turns out to be harmless, but I just wanted to point it out because it looks like it may be unintentional.

Reply:
Also noticed this and was a bit confused. In the case where `remove` is true, `cache_descendants` never gets added to the `descendants_by_txid` map, which is fine. In the case where `remove` is false, then `txid` should be in `entries_by_txid`, so it will get added to `cache_descendants`, along with all of `descendants`.

So what gets added to `descendants_by_txid` is `txidA: [txidA, txidB, txidC...]`, basically the parent + all of its children, which seems incorrect?

Reply:
@josibake 
> ... which seems incorrect?

No, I think it's a convention in the strange Land of Mempool that a transaction's ancestor and descendant lists both include the transaction itself. This relates directly to my [comment](https://github.com/bitcoin/bitcoin/pull/26152#discussion_r1050404621) elsewhere that the `emplace_back()` should be removed; as it's currently written, the list would be: `txidA: [txidA, txidA, txidB, txidC...]`. This later causes a major accounting problem!

Reply:
Yes, it would be more correct to gate the `emplace_back()` on `!remove`.